File: mobile/lib/utils/expense_reason_matcher.dart
/// Utility for matching expense references in salary withdrawal reason text.
///
/// Uses a regex with a negative lookahead to prevent `exp_1` from matching
/// `exp_10`, `exp_100`, etc.
bool matchesExpenseRef(String? reason, int expenseId) {
  if (reason == null) return false;
  return RegExp('exp_' + expenseId.toString() + r'(?!\d)').hasMatch(reason);
}

File: mobile/android/app/src/main/kotlin/com/marina/marina/domain/util/ExpenseReasonMatcher.kt
package com.marina.marina.domain.util

/**
 * مطابقة مرجع المصروف في نص reason لسحب الراتب — مرآة 1:1 لملف
 * Flutter `mobile/lib/utils/expense_reason_matcher.dart`.
 *
 * يستخدم تعبيراً نمطياً مع negative lookahead يمنع `exp_1` من مطابقة
 * `exp_10` أو `exp_100` — العقد الذي تعتمد عليه إزالة تكرار تقرير
 * المصروفات بين جدولي expenses و salary_withdrawals.
 */
object ExpenseReasonMatcher {

    /** true عندما يحمل reason مرجع exp_<expenseId> الحرفي (وليس بامتداد رقمي أطول). */
    fun matchesExpenseRef(reason: String?, expenseId: Long): Boolean {
        if (reason == null) return false
        return Regex("exp_$expenseId(?!\\d)").containsMatchIn(reason)
    }
}
